src/Client.php

```diff
@@ -275,4 +275,151 @@ public function getAsset($assetId) {
     return (string) $response->getBody();
   }
 
+  /**
+   * Gets presigned url from AWS S3.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam.
+   *
+   * @return mixed
+   *   Presigned url needed for next step + PID.
+   */
+  public function getPresignUrl(array $file_data) {
+    $this->checkAuth();
+
+    try {
+      $response = $this->client->request(
+        "GET",
+        $this->baseUrl . '/ws/awss3/generateupload',
+        [
+          'headers' => $this->getDefaultHeaders(),
+          'query' => $file_data,
+        ]
+      );
+      return json_decode($response->getBody());
+    }
+    catch (ClientException $e) {
+      dump($e->getMessage());
+      return FALSE;
+    }
+  }
+
+  /**
+   * Uploads file to Webdam AWS S3.
+   *
+   * @param mixed $presignedUrl
+   *   The presigned URL we got in previous step from AWS.
+   * @param array $file_data
+   *   The file data required by Webdam.
+   *
+   * @return array
+   *   Response Status 100 / 200
+   */
+  public function uploadPresigned($presignedUrl, array $file_data) {
+    $this->checkAuth();
+
+    $file = fopen($file_data['file_uri'], 'r');
```

Comment:
Instead of having the client open the file directly, let's just pass the contents of the file in a variable (make it the caller's responsibility to load the data - that simplifies our tests)